### PR TITLE
feat: update WalletConnectConnector for v2

### PR DESCRIPTION
**Pull request by ganchoradkov**
## Description

Add WalletConnect v2 support to the existing WC Connector.

## Additional Information
WalletConnect v1 is to be depreciated on March 1, 2023.
This v2 connector will allow wagmi users to have enough time to prepare and migrate to WalletConnect v2 which offers significantly better developer & user experience

- [x] I read the [contributing docs](/wagmi-dev/wagmi/blob/main/.github/CONTRIBUTING.md) (if this is your first contribution)

Your ENS/address:

### CODE REVIEW

File: packages/core/package.json [hunk 1]
```diff
@@ -84,6 +84,8 @@
   "peerDependencies": {
     "@coinbase/wallet-sdk": ">=3.3.0",
     "@walletconnect/ethereum-provider": ">=1.7.5",
+    "@walletconnect/qrcode-modal": ">=1.8.0",
+    "@walletconnect/universal-provider": ">=2.1.0",
```
Comment (tmm) [inline packages/core/package.json:88]: Added range in

File: packages/core/src/connectors/walletConnect.ts [hunk 2]
```diff
@@ -92,21 +188,31 @@ export class WalletConnectConnector extends Connector<
...
+    }
   }
 
   async getAccount() {
     const provider = await this.getProvider()
-    const accounts = provider.accounts
+    let accounts
+    if (this.version === '1')
+      accounts = (provider as WalletConnectProvider).accounts
+    else accounts = await provider.enable() // TODO: Should we use `.request({ method: 'eth_accounts' })` instead?
```
Comment (tmm) [inline packages/core/src/connectors/walletConnect.ts:205]: This returns an empty array when calling `isAuthorized` so auto connecting is not working.

File: packages/core/src/connectors/walletConnect.ts [hunk 3]
```diff
@@ -206,4 +325,8 @@ export class WalletConnectConnector extends Connector<
   protected onDisconnect = () => {
     this.emit('disconnect')
   }
+
+  protected onDisplayUri = (uri: string) => {
+    this.emit('message', { type: 'display_uri', data: uri })
+  }
```
Comment (tmm) [inline packages/core/src/connectors/walletConnect.ts:331]: Decoupled `Modal` from this

File: packages/core/src/connectors/walletConnect.ts [hunk 4]
```diff
@@ -56,6 +73,58 @@ export class WalletConnectConnector extends Connector<
...
+                new Promise((_res, reject) =>
+                  provider.on('display_uri', async (uri: string) => {
+                    const { default: Modal } = await import(
+                      '@walletconnect/qrcode-modal'
+                    )
+                    Modal.open(uri, () => reject(new Error('user rejected')))
+                  }),
+                ),
+              ]
+            : []),
```
Comment (tmm) [inline packages/core/src/connectors/walletConnect.ts:118]: Need this block so when user closes out of QR Code Modal, "rejected request" error bubbles up to user.

File: packages/core/src/connectors/walletConnect.ts [hunk 5]
```diff
@@ -56,6 +73,58 @@ export class WalletConnectConnector extends Connector<
...
+        ])
+
+        // If execution reaches here, connection was successful and we can close modal.
+        if (this.options.qrcode) {
+          const { default: Modal } = await import('@walletconnect/qrcode-modal')
+          Modal.close()
+        }
```
Comment (tmm) [inline packages/core/src/connectors/walletConnect.ts:125]: This assumption feels a little shaky. Is this the best way to know if `provider.connect` succeeded?

File: packages/core/src/connectors/walletConnect.ts (hunk 2 above)
Comment (ganchoradkov) [reply packages/core/src/connectors/walletConnect.ts:208]: Updated the `universal-provider` to `2.1.4` where this issue is resolved. Now returns correct the address from its persistence storage after reload

File: packages/core/src/connectors/walletConnect.ts (hunk 5 above)
Comment (ganchoradkov) [reply packages/core/src/connectors/walletConnect.ts:128]: I think it should be fine. If the connection request is declined by the user an exception is thrown and is being captured correctly `catch` block

File: packages/core/src/connectors/walletConnect.ts [hunk 6]
```diff
@@ -9,36 +11,49 @@ import type { Chain } from '../types'
...
+  qrcode?: boolean
+} & (
+  | ({
+      version?: '1'
+    } & ConstructorParameters<typeof WalletConnectProvider>[0])
+  | ({
+      /**
+       * Project ID associated with [WalletConnect account](https://cloud.walletconnect.com)
+       */
+      projectId: NonNullable<UniversalProviderOpts['projectId']>
```
Comment (seendsouza) [inline packages/core/src/connectors/walletConnect.ts:30]: Isn't it possible to have no projectId if using your own relayer for walletconnect v2?

**Comment by jxom** [comment]
Closing in favor of https://github.com/wagmi-dev/references/pull/19

**Pull request closed without merging**